diff --git a/src/pulse/fork-detect.c b/src/pulse/fork-detect.c
index 9299304..1e6cf4f 100755
--- a/src/pulse/fork-detect.c
+++ b/src/pulse/fork-detect.c
@@ -37,7 +37,7 @@ int pa_detect_fork(void) {
      * however have to deal with this cleanly, so we try to detect the
      * forks making sure all our calls fail cleanly after the fork. */
 
-    pa_assert_cc(sizeof(pa_atomic_t) >= sizeof(pid_t));
+    //pa_assert_cc(sizeof(pa_atomic_t) >= sizeof(pid_t));
 
     for (;;) {
         pid_t stored_pid = (pid_t) pa_atomic_load(&pid);

diff --git a/src/pulsecore/arpa-inet.c b/src/pulsecore/arpa-inet.c
index fb34299..6110c1c 100755
--- a/src/pulsecore/arpa-inet.c
+++ b/src/pulsecore/arpa-inet.c
@@ -16,7 +16,7 @@
   You should have received a copy of the GNU Lesser General Public
   License along with PulseAudio; if not, see <http://www.gnu.org/licenses/>.
 ***/
-
+#ifndef _WIN32
 #ifdef HAVE_CONFIG_H
 #include <config.h>
 #endif
@@ -104,3 +104,4 @@ int inet_pton(int af, const char *src, void *dst) {
 }
 
 #endif
+#endif
