diff --git a/src/cmdutils.h b/src/cmdutils.h
index 85b468f..f6a3d23 100755
--- a/src/cmdutils.h
+++ b/src/cmdutils.h
@@ -1,3 +1,4 @@
+#include "ffmpeg_tls.h"
 /*
  * Various utilities for command line tools
  * copyright (c) 2003 Fabrice Bellard
@@ -37,12 +38,12 @@
 /**
  * program name, defined by the program for show_version().
  */
-extern const char program_name[];
+extern FFMPEG_THREAD_LOCAL const char program_name[];
 
 /**
  * program birth year, defined by the program for show_banner()
  */
-extern const int program_birth_year;
+extern FFMPEG_THREAD_LOCAL const int program_birth_year;
 
 extern AVDictionary *sws_dict;
 extern AVDictionary *swr_opts;
@@ -283,10 +284,22 @@ void show_help_options(const OptionDef *options, const char *msg, int req_flags,
 void show_help_children(const AVClass *class, int flags);
 
 /**
- * Per-fftool specific help handler. Implemented in each
- * fftool, called by show_help().
+ * FFmpeg specific help handler. Implemented in each
+ * FFmpeg, called by show_help().
  */
-void show_help_default(const char *opt, const char *arg);
+void ffmpeg_show_help_default(const char *opt, const char *arg);
+
+/**
+ * FFprobe specific help handler. Implemented in each
+ * FFprobe, called by show_help().
+ */
+void ffprobe_show_help_default(const char *opt, const char *arg);
+
+/**
+ * FFplay specific help handler. Implemented in each
+ * FFplay, called by show_help().
+ */
+void ffplay_show_help_default(const char *opt, const char *arg);
 
 /**
  * Parse the command line arguments.
