diff --git a/src/ffmpeg_sched.c b/src/ffmpeg_sched.c
index 3180367..e824da9 100755
--- a/src/ffmpeg_sched.c
+++ b/src/ffmpeg_sched.c
@@ -2524,6 +2524,8 @@ static int task_cleanup(Scheduler *sch, SchedulerNode node)
     }
 }
 
+extern void ffmpeg_tls_init_options(void);
+
 static void *task_wrapper(void *arg)
 {
     SchTask  *task = arg;
@@ -2531,6 +2533,9 @@ static void *task_wrapper(void *arg)
     int ret;
     int err = 0;
 
+    // FFmpegKit: Initialize TLS options for the sub-thread
+    ffmpeg_tls_init_options();
+
     ret = task->func(task->func_arg);
     if (ret < 0)
         av_log(task->func_arg, AV_LOG_ERROR,
@@ -2565,10 +2570,17 @@ static int task_stop(Scheduler *sch, SchTask *task)
         return task_cleanup(sch, task->node);
 
     ret = pthread_join(task->thread, &thread_ret);
+
+#ifdef _WIN32
+    if (ret == ESRCH) {
+        task->thread_running = 0;
+        return task_cleanup(sch, task->node);
+    }
+#endif
+
     av_assert0(ret == 0);
 
     task->thread_running = 0;
-
     return (intptr_t)thread_ret;
 }
 
