diff --git a/src/opt_common.c b/src/opt_common.c
index 2c1f956..a00786d 100755
--- a/src/opt_common.c
+++ b/src/opt_common.c
@@ -71,15 +71,17 @@ enum show_muxdemuxers {
 static FILE *report_file;
 static int report_file_level = AV_LOG_DEBUG;
 
+extern void FFMPEG_THREAD_LOCAL (*show_help_default_func)(const char *opt, const char *arg);
+
 int show_license(void *optctx, const char *opt, const char *arg)
 {
 #if CONFIG_NONFREE
-    printf(
+    av_log(NULL, AV_LOG_INFO, 
     "This version of %s has nonfree parts compiled in.\n"
     "Therefore it is not legally redistributable.\n",
     program_name );
 #elif CONFIG_GPLV3
-    printf(
+    av_log(NULL, AV_LOG_INFO, 
     "%s is free software; you can redistribute it and/or modify\n"
     "it under the terms of the GNU General Public License as published by\n"
     "the Free Software Foundation; either version 3 of the License, or\n"
@@ -94,7 +96,7 @@ int show_license(void *optctx, const char *opt, const char *arg)
     "along with %s.  If not, see <http://www.gnu.org/licenses/>.\n",
     program_name, program_name, program_name );
 #elif CONFIG_GPL
-    printf(
+    av_log(NULL, AV_LOG_INFO, 
     "%s is free software; you can redistribute it and/or modify\n"
     "it under the terms of the GNU General Public License as published by\n"
     "the Free Software Foundation; either version 2 of the License, or\n"
@@ -110,7 +112,7 @@ int show_license(void *optctx, const char *opt, const char *arg)
     "Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA\n",
     program_name, program_name, program_name );
 #elif CONFIG_LGPLV3
-    printf(
+    av_log(NULL, AV_LOG_INFO, 
     "%s is free software; you can redistribute it and/or modify\n"
     "it under the terms of the GNU Lesser General Public License as published by\n"
     "the Free Software Foundation; either version 3 of the License, or\n"
@@ -125,7 +127,7 @@ int show_license(void *optctx, const char *opt, const char *arg)
     "along with %s.  If not, see <http://www.gnu.org/licenses/>.\n",
     program_name, program_name, program_name );
 #else
-    printf(
+    av_log(NULL, AV_LOG_INFO, 
     "%s is free software; you can redistribute it and/or\n"
     "modify it under the terms of the GNU Lesser General Public\n"
     "License as published by the Free Software Foundation; either\n"
@@ -245,7 +247,6 @@ void show_banner(int argc, char **argv, const OptionDef *options)
 
 int show_version(void *optctx, const char *opt, const char *arg)
 {
-    av_log_set_callback(log_callback_help);
     print_program_info (SHOW_COPYRIGHT, AV_LOG_INFO);
     print_all_libs_info(SHOW_VERSION, AV_LOG_INFO);
 
@@ -254,7 +255,6 @@ int show_version(void *optctx, const char *opt, const char *arg)
 
 int show_buildconf(void *optctx, const char *opt, const char *arg)
 {
-    av_log_set_callback(log_callback_help);
     print_buildconf      (INDENT|0, AV_LOG_INFO);
 
     return 0;
@@ -291,58 +291,58 @@ static void print_codec(const AVCodec *c)
     int encoder = av_codec_is_encoder(c);
     AVBPrint desc;
 
-    printf("%s %s [%s]:\n", encoder ? "Encoder" : "Decoder", c->name,
+    av_log(NULL, AV_LOG_INFO, "%s %s [%s]:\n", encoder ? "Encoder" : "Decoder", c->name,
            c->long_name ? c->long_name : "");
 
-    printf("    General capabilities: ");
+    av_log(NULL, AV_LOG_INFO, "    General capabilities: ");
     if (c->capabilities & AV_CODEC_CAP_DRAW_HORIZ_BAND)
-        printf("horizband ");
+        av_log(NULL, AV_LOG_INFO, "horizband ");
     if (c->capabilities & AV_CODEC_CAP_DR1)
-        printf("dr1 ");
+        av_log(NULL, AV_LOG_INFO, "dr1 ");
     if (c->capabilities & AV_CODEC_CAP_DELAY)
-        printf("delay ");
+        av_log(NULL, AV_LOG_INFO, "delay ");
     if (c->capabilities & AV_CODEC_CAP_SMALL_LAST_FRAME)
-        printf("small ");
+        av_log(NULL, AV_LOG_INFO, "small ");
     if (c->capabilities & AV_CODEC_CAP_EXPERIMENTAL)
-        printf("exp ");
+        av_log(NULL, AV_LOG_INFO, "exp ");
     if (c->capabilities & AV_CODEC_CAP_CHANNEL_CONF)
-        printf("chconf ");
+        av_log(NULL, AV_LOG_INFO, "chconf ");
     if (c->capabilities & AV_CODEC_CAP_PARAM_CHANGE)
-        printf("paramchange ");
+        av_log(NULL, AV_LOG_INFO, "paramchange ");
     if (c->capabilities & AV_CODEC_CAP_VARIABLE_FRAME_SIZE)
-        printf("variable ");
+        av_log(NULL, AV_LOG_INFO, "variable ");
     if (c->capabilities & (AV_CODEC_CAP_FRAME_THREADS |
                            AV_CODEC_CAP_SLICE_THREADS |
                            AV_CODEC_CAP_OTHER_THREADS))
-        printf("threads ");
+        av_log(NULL, AV_LOG_INFO, "threads ");
     if (c->capabilities & AV_CODEC_CAP_AVOID_PROBING)
-        printf("avoidprobe ");
+        av_log(NULL, AV_LOG_INFO, "avoidprobe ");
     if (c->capabilities & AV_CODEC_CAP_HARDWARE)
-        printf("hardware ");
+        av_log(NULL, AV_LOG_INFO, "hardware ");
     if (c->capabilities & AV_CODEC_CAP_HYBRID)
-        printf("hybrid ");
+        av_log(NULL, AV_LOG_INFO, "hybrid ");
     if (!c->capabilities)
-        printf("none");
-    printf("\n");
+        av_log(NULL, AV_LOG_INFO, "none");
+    av_log(NULL, AV_LOG_INFO, "\n");
 
     if (c->type == AVMEDIA_TYPE_VIDEO ||
         c->type == AVMEDIA_TYPE_AUDIO) {
-        printf("    Threading capabilities: ");
+        av_log(NULL, AV_LOG_INFO, "    Threading capabilities: ");
         switch (c->capabilities & (AV_CODEC_CAP_FRAME_THREADS |
                                    AV_CODEC_CAP_SLICE_THREADS |
                                    AV_CODEC_CAP_OTHER_THREADS)) {
         case AV_CODEC_CAP_FRAME_THREADS |
-             AV_CODEC_CAP_SLICE_THREADS: printf("frame and slice"); break;
-        case AV_CODEC_CAP_FRAME_THREADS: printf("frame");           break;
-        case AV_CODEC_CAP_SLICE_THREADS: printf("slice");           break;
-        case AV_CODEC_CAP_OTHER_THREADS: printf("other");           break;
-        default:                         printf("none");            break;
+             AV_CODEC_CAP_SLICE_THREADS: av_log(NULL, AV_LOG_INFO, "frame and slice"); break;
+        case AV_CODEC_CAP_FRAME_THREADS: av_log(NULL, AV_LOG_INFO, "frame");           break;
+        case AV_CODEC_CAP_SLICE_THREADS: av_log(NULL, AV_LOG_INFO, "slice");           break;
+        case AV_CODEC_CAP_OTHER_THREADS: av_log(NULL, AV_LOG_INFO, "other");           break;
+        default:                         av_log(NULL, AV_LOG_INFO, "none");            break;
         }
-        printf("\n");
+        av_log(NULL, AV_LOG_INFO, "\n");
     }
 
     if (avcodec_get_hw_config(c, 0)) {
-        printf("    Supported hardware devices: ");
+        av_log(NULL, AV_LOG_INFO, "    Supported hardware devices: ");
         for (int i = 0;; i++) {
             const AVCodecHWConfig *config = avcodec_get_hw_config(c, i);
             const char *name;
@@ -350,9 +350,9 @@ static void print_codec(const AVCodec *c)
                 break;
             name = av_hwdevice_get_type_name(config->device_type);
             if (name)
-                printf("%s ", name);
+                av_log(NULL, AV_LOG_INFO, "%s ", name);
         }
-        printf("\n");
+        av_log(NULL, AV_LOG_INFO, "\n");
     }
 
     PRINT_CODEC_SUPPORTED(c, AV_CODEC_CONFIG_FRAME_RATE, AVRational, "framerates",
@@ -434,10 +434,10 @@ static void show_help_demuxer(const char *name)
         return;
     }
 
-    printf("Demuxer %s [%s]:\n", fmt->name, fmt->long_name);
+    av_log(NULL, AV_LOG_INFO, "Demuxer %s [%s]:", fmt->name, fmt->long_name);
 
     if (fmt->extensions)
-        printf("    Common extensions: %s.\n", fmt->extensions);
+        av_log(NULL, AV_LOG_INFO, "    Common extensions: %s.", fmt->extensions);
 
     if (fmt->priv_class)
         show_help_children(fmt->priv_class, AV_OPT_FLAG_DECODING_PARAM);
@@ -471,23 +471,23 @@ static void show_help_muxer(const char *name)
         return;
     }
 
-    printf("Muxer %s [%s]:\n", fmt->name, fmt->long_name);
+    av_log(NULL, AV_LOG_INFO, "Muxer %s [%s]:", fmt->name, fmt->long_name);
 
     if (fmt->extensions)
-        printf("    Common extensions: %s.\n", fmt->extensions);
+        av_log(NULL, AV_LOG_INFO, "    Common extensions: %s.", fmt->extensions);
     if (fmt->mime_type)
-        printf("    Mime type: %s.\n", fmt->mime_type);
+        av_log(NULL, AV_LOG_INFO, "    Mime type: %s.", fmt->mime_type);
     if (fmt->video_codec != AV_CODEC_ID_NONE &&
         (desc = avcodec_descriptor_get(fmt->video_codec))) {
-        printf("    Default video codec: %s.\n", desc->name);
+        av_log(NULL, AV_LOG_INFO, "    Default video codec: %s.", desc->name);
     }
     if (fmt->audio_codec != AV_CODEC_ID_NONE &&
         (desc = avcodec_descriptor_get(fmt->audio_codec))) {
-        printf("    Default audio codec: %s.\n", desc->name);
+        av_log(NULL, AV_LOG_INFO, "    Default audio codec: %s.", desc->name);
     }
     if (fmt->subtitle_codec != AV_CODEC_ID_NONE &&
         (desc = avcodec_descriptor_get(fmt->subtitle_codec))) {
-        printf("    Default subtitle codec: %s.\n", desc->name);
+        av_log(NULL, AV_LOG_INFO, "    Default subtitle codec: %s.", desc->name);
     }
 
     if (fmt->priv_class)
@@ -509,40 +509,40 @@ static void show_help_filter(const char *name)
         return;
     }
 
-    printf("Filter %s\n", f->name);
+    av_log(NULL, AV_LOG_INFO, "Filter %s", f->name);
     if (f->description)
-        printf("  %s\n", f->description);
+        av_log(NULL, AV_LOG_INFO, "  %s", f->description);
 
     if (f->flags & AVFILTER_FLAG_SLICE_THREADS)
-        printf("    slice threading supported\n");
+        av_log(NULL, AV_LOG_INFO, "    slice threading supported");
 
-    printf("    Inputs:\n");
+    av_log(NULL, AV_LOG_INFO, "    Inputs:");
     count = avfilter_filter_pad_count(f, 0);
     for (i = 0; i < count; i++) {
-        printf("       #%d: %s (%s)\n", i, avfilter_pad_get_name(f->inputs, i),
+        av_log(NULL, AV_LOG_INFO, "       #%d: %s (%s)", i, avfilter_pad_get_name(f->inputs, i),
                av_get_media_type_string(avfilter_pad_get_type(f->inputs, i)));
     }
     if (f->flags & AVFILTER_FLAG_DYNAMIC_INPUTS)
-        printf("        dynamic (depending on the options)\n");
+        av_log(NULL, AV_LOG_INFO, "        dynamic (depending on the options)");
     else if (!count)
-        printf("        none (source filter)\n");
+        av_log(NULL, AV_LOG_INFO, "        none (source filter)");
 
-    printf("    Outputs:\n");
+    av_log(NULL, AV_LOG_INFO, "    Outputs:");
     count = avfilter_filter_pad_count(f, 1);
     for (i = 0; i < count; i++) {
-        printf("       #%d: %s (%s)\n", i, avfilter_pad_get_name(f->outputs, i),
+        av_log(NULL, AV_LOG_INFO, "       #%d: %s (%s)", i, avfilter_pad_get_name(f->outputs, i),
                av_get_media_type_string(avfilter_pad_get_type(f->outputs, i)));
     }
     if (f->flags & AVFILTER_FLAG_DYNAMIC_OUTPUTS)
-        printf("        dynamic (depending on the options)\n");
+        av_log(NULL, AV_LOG_INFO, "        dynamic (depending on the options)");
     else if (!count)
-        printf("        none (sink filter)\n");
+        av_log(NULL, AV_LOG_INFO, "        none (sink filter)");
 
     if (f->priv_class)
         show_help_children(f->priv_class, AV_OPT_FLAG_VIDEO_PARAM | AV_OPT_FLAG_FILTERING_PARAM |
                                           AV_OPT_FLAG_AUDIO_PARAM);
     if (f->flags & AVFILTER_FLAG_SUPPORT_TIMELINE)
-        printf("This filter has support for timeline through the 'enable' option.\n");
+        av_log(NULL, AV_LOG_INFO, "This filter has support for timeline through the 'enable' option.");
 #else
     av_log(NULL, AV_LOG_ERROR, "Build without libavfilter; "
            "can not to satisfy request\n");
@@ -562,15 +562,15 @@ static void show_help_bsf(const char *name)
         return;
     }
 
-    printf("Bit stream filter %s\n", bsf->name);
+    av_log(NULL, AV_LOG_INFO, "Bit stream filter %s", bsf->name);
     if (bsf->codec_ids) {
         const enum AVCodecID *id = bsf->codec_ids;
-        printf("    Supported codecs:");
+        av_log(NULL, AV_LOG_INFO, "    Supported codecs:");
         while (*id != AV_CODEC_ID_NONE) {
-            printf(" %s", avcodec_descriptor_get(*id)->name);
+            av_log(NULL, AV_LOG_INFO, " %s", avcodec_descriptor_get(*id)->name);
             id++;
         }
-        printf("\n");
+        av_log(NULL, AV_LOG_INFO, "\n");
     }
     if (bsf->priv_class)
         show_help_children(bsf->priv_class, AV_OPT_FLAG_BSF_PARAM);
@@ -579,7 +579,6 @@ static void show_help_bsf(const char *name)
 int show_help(void *optctx, const char *opt, const char *arg)
 {
     char *topic, *par;
-    av_log_set_callback(log_callback_help);
 
     topic = av_strdup(arg ? arg : "");
     if (!topic)
@@ -589,7 +588,8 @@ int show_help(void *optctx, const char *opt, const char *arg)
         *par++ = 0;
 
     if (!*topic) {
-        show_help_default(topic, par);
+        if (show_help_default_func)
+            show_help_default_func(topic, par);
     } else if (!strcmp(topic, "decoder")) {
         show_help_codec(par, 0);
     } else if (!strcmp(topic, "encoder")) {
@@ -607,7 +607,8 @@ int show_help(void *optctx, const char *opt, const char *arg)
     } else if (!strcmp(topic, "bsf")) {
         show_help_bsf(par);
     } else {
-        show_help_default(topic, par);
+        if (show_help_default_func)
+            show_help_default_func(topic, par);
     }
 
     av_freep(&topic);
