diff --git a/src/ffplay.c b/src/ffplay.c
index 2a572fc..cf4b33f 100755
--- a/src/ffplay.c
+++ b/src/ffplay.c
@@ -58,8 +58,8 @@
 #include "ffplay_renderer.h"
 #include "opt_common.h"
 
-const char program_name[] = "ffplay";
-const int program_birth_year = 2003;
+//const char program_name[] = "ffplay";
+//const int program_birth_year = 2003;
 
 #define MAX_QUEUE_SIZE (15 * 1024 * 1024)
 #define MIN_FRAMES 25
@@ -83,6 +83,8 @@ const int program_birth_year = 2003;
 /* no AV correction is done if too big error */
 #define AV_NOSYNC_THRESHOLD 10.0
 
+extern void (*ffplay_on_show_help)(void);
+
 /* maximum audio speed change to get correct sync */
 #define SAMPLE_CORRECTION_PERCENT_MAX 10
 
@@ -299,6 +301,11 @@ typedef struct VideoState {
     int last_video_stream, last_audio_stream, last_subtitle_stream;
 
     SDL_cond *continue_read_thread;
+    int autoexit;
+    int loop;
+    int64_t duration;
+    int64_t start_time;
+    int infinite_buffer;
 } VideoState;
 
 /* options specified by the user */
@@ -1222,6 +1229,7 @@ static void stream_component_close(VideoState *is, int stream_index)
         av_freep(&is->audio_buf1);
         is->audio_buf1_size = 0;
         is->audio_buf = NULL;
+        avfilter_graph_free(&is->agraph);
 
         if (is->rdft) {
             av_tx_uninit(&is->rdft);
@@ -1322,12 +1330,13 @@ static void do_exit(VideoState *is)
         printf("\n");
     SDL_Quit();
     av_log(NULL, AV_LOG_QUIET, "%s", "");
-    exit(0);
+    //exit(0);
 }
 
 static void sigterm_handler(int sig)
 {
-    exit(123);
+    //exit(123);
+
 }
 
 static void set_default_window_size(int width, int height, AVRational sar)
@@ -2787,6 +2796,10 @@ static int stream_component_open(VideoState *is, int stream_index)
 
 fail:
     avcodec_free_context(&avctx);
+    if (is->agraph != NULL) {
+        avfilter_graph_free(&is->agraph);
+    }
+
 out:
     av_channel_layout_uninit(&ch_layout);
     av_dict_free(&opts);
@@ -2846,7 +2859,7 @@ static int read_thread(void *arg)
 
     memset(st_index, -1, sizeof(st_index));
     is->eof = 0;
-
+    
     pkt = av_packet_alloc();
     if (!pkt) {
         av_log(NULL, AV_LOG_FATAL, "Could not allocate packet.\n");
@@ -2920,21 +2933,22 @@ static int read_thread(void *arg)
     is->max_frame_duration = (ic->iformat->flags & AVFMT_TS_DISCONT) ? 10.0 : 3600.0;
 
     if (!window_title && (t = av_dict_get(ic->metadata, "title", NULL, 0)))
-        window_title = av_asprintf("%s - %s", t->value, input_filename);
+        window_title = av_asprintf("%s - %s", t->value, is->filename);
 
     /* if seeking requested, we execute it */
-    if (start_time != AV_NOPTS_VALUE) {
+    if (is->start_time != AV_NOPTS_VALUE) {
         int64_t timestamp;
 
-        timestamp = start_time;
+        timestamp = is->start_time;
         /* add the stream start time */
-        if (ic->start_time != AV_NOPTS_VALUE)
-            timestamp += ic->start_time;
+        if (is->start_time != AV_NOPTS_VALUE && is->start_time > 0) {
+        int64_t seek_pos = is->start_time;
         ret = avformat_seek_file(ic, -1, INT64_MIN, timestamp, INT64_MAX, 0);
-        if (ret < 0) {
-            av_log(NULL, AV_LOG_WARNING, "%s: could not seek to position %0.3f\n",
-                    is->filename, (double)timestamp / AV_TIME_BASE);
-        }
+        if (is->duration != AV_NOPTS_VALUE) {
+        av_log(NULL, AV_LOG_INFO, "Duration: %02"PRId64":%02"PRId64":%02"PRId64".%02"PRId64"\n",
+               is->duration / 3600, (is->duration / 60) % 60, is->duration % 60, (is->duration * 100 / AV_TIME_BASE) % 100);
+    }
+    }
     }
 
     is->realtime = is_realtime(ic);
@@ -2976,7 +2990,7 @@ static int read_thread(void *arg)
                                  st_index[AVMEDIA_TYPE_VIDEO]),
                                 NULL, 0);
 
-    is->show_mode = show_mode;
+    // is->show_mode = show_mode; // Removed to respect initialized value
     if (st_index[AVMEDIA_TYPE_VIDEO] >= 0) {
         AVStream *st = ic->streams[st_index[AVMEDIA_TYPE_VIDEO]];
         AVCodecParameters *codecpar = st->codecpar;
@@ -3008,8 +3022,8 @@ static int read_thread(void *arg)
         goto fail;
     }
 
-    if (infinite_buffer < 0 && is->realtime)
-        infinite_buffer = 1;
+    if (is->infinite_buffer < 0 && is->realtime)
+        is->infinite_buffer = 1;
 
     for (;;) {
         if (is->abort_request)
@@ -3024,7 +3038,7 @@ static int read_thread(void *arg)
 #if CONFIG_RTSP_DEMUXER || CONFIG_MMSH_PROTOCOL
         if (is->paused &&
                 (!strcmp(ic->iformat->name, "rtsp") ||
-                 (ic->pb && !strncmp(input_filename, "mmsh:", 5)))) {
+                 (ic->pb && !strncmp(is->filename, "mmsh:", 5)))) {
             /* wait 10 ms to avoid trying to get another packet */
             /* XXX: horrible */
             SDL_Delay(10);
@@ -3072,7 +3086,7 @@ static int read_thread(void *arg)
         }
 
         /* if the queue are full, no need to read more */
-        if (infinite_buffer<1 &&
+        if (is->infinite_buffer < 1 &&
               (is->audioq.size + is->videoq.size + is->subtitleq.size > MAX_QUEUE_SIZE
             || (stream_has_enough_packets(is->audio_st, is->audio_stream, &is->audioq) &&
                 stream_has_enough_packets(is->video_st, is->video_stream, &is->videoq) &&
@@ -3086,9 +3100,9 @@ static int read_thread(void *arg)
         if (!is->paused &&
             (!is->audio_st || (is->auddec.finished == is->audioq.serial && frame_queue_nb_remaining(&is->sampq) == 0)) &&
             (!is->video_st || (is->viddec.finished == is->videoq.serial && frame_queue_nb_remaining(&is->pictq) == 0))) {
-            if (loop != 1 && (!loop || --loop)) {
-                stream_seek(is, start_time != AV_NOPTS_VALUE ? start_time : 0, 0, 0);
-            } else if (autoexit) {
+            if (is->loop != 1 && (!is->loop || --is->loop)) {
+            stream_seek(is, is->start_time != AV_NOPTS_VALUE ? is->start_time : 0, 0, 0);
+        } else if (is->autoexit) {
                 ret = AVERROR_EOF;
                 goto fail;
             }
@@ -3105,7 +3119,7 @@ static int read_thread(void *arg)
                 is->eof = 1;
             }
             if (ic->pb && ic->pb->error) {
-                if (autoexit)
+                if (is->autoexit)
                     goto fail;
                 else
                     break;
@@ -3120,11 +3134,11 @@ static int read_thread(void *arg)
         /* check if packet is in play range specified by user, then queue, otherwise discard */
         stream_start_time = ic->streams[pkt->stream_index]->start_time;
         pkt_ts = pkt->pts == AV_NOPTS_VALUE ? pkt->dts : pkt->pts;
-        pkt_in_play_range = duration == AV_NOPTS_VALUE ||
+        pkt_in_play_range = is->duration == AV_NOPTS_VALUE ||
                 (pkt_ts - (stream_start_time != AV_NOPTS_VALUE ? stream_start_time : 0)) *
                 av_q2d(ic->streams[pkt->stream_index]->time_base) -
-                (double)(start_time != AV_NOPTS_VALUE ? start_time : 0) / 1000000
-                <= ((double)duration / 1000000);
+                (double)(is->start_time != AV_NOPTS_VALUE ? is->start_time : 0) / 1000000
+                <= ((double)is->duration / 1000000);
         if (pkt->stream_index == is->audio_stream && pkt_in_play_range) {
             packet_queue_put(&is->audioq, pkt);
         } else if (pkt->stream_index == is->video_stream && pkt_in_play_range
@@ -3171,6 +3185,14 @@ static VideoState *stream_open(const char *filename,
     is->iformat = iformat;
     is->ytop    = 0;
     is->xleft   = 0;
+    
+    /* init TLS-copied state */
+    is->loop = loop;
+    is->duration = duration;
+    is->start_time = start_time;
+    is->infinite_buffer = infinite_buffer;
+    is->autoexit = autoexit;
+    is->show_mode = show_mode;
 
     /* start video display */
     if (frame_queue_init(&is->pictq, &is->videoq, VIDEO_PICTURE_QUEUE_SIZE, 1) < 0)
@@ -3194,6 +3216,7 @@ static VideoState *stream_open(const char *filename,
     init_clock(&is->audclk, &is->audioq.serial);
     init_clock(&is->extclk, &is->extclk.serial);
     is->audio_clock_serial = -1;
+
     if (startup_volume < 0)
         av_log(NULL, AV_LOG_WARNING, "-volume=%d < 0, setting to 0\n", startup_volume);
     if (startup_volume > 100)
@@ -3594,7 +3617,7 @@ static int opt_sync(void *optctx, const char *opt, const char *arg)
         av_sync_type = AV_SYNC_EXTERNAL_CLOCK;
     else {
         av_log(NULL, AV_LOG_ERROR, "Unknown value for %s: %s\n", opt, arg);
-        exit(1);
+        //exit(1);
     }
     return 0;
 }
@@ -3722,12 +3745,13 @@ static void show_usage(void)
     av_log(NULL, AV_LOG_INFO, "\n");
 }
 
-void show_help_default(const char *opt, const char *arg)
+void ffplay_show_help_default(const char *opt, const char *arg)
 {
     av_log_set_callback(log_callback_help);
     show_usage();
     show_help_options(options, "Main options:", 0, OPT_EXPERT);
     show_help_options(options, "Advanced options:", OPT_EXPERT, 0);
+    if (ffplay_on_show_help) ffplay_on_show_help();
     printf("\n");
     show_help_children(avcodec_get_class(), AV_OPT_FLAG_DECODING_PARAM);
     show_help_children(avformat_get_class(), AV_OPT_FLAG_DECODING_PARAM);
@@ -3753,8 +3777,10 @@ void show_help_default(const char *opt, const char *arg)
            );
 }
 
+void (*ffplay_on_show_help)(void) = NULL;
+
 /* Called from the main */
-int main(int argc, char **argv)
+VideoState *ffplay_init_internal(int argc, char **argv)
 {
     int flags, ret;
     VideoState *is;
@@ -3777,14 +3803,14 @@ int main(int argc, char **argv)
 
     ret = parse_options(NULL, argc, argv, options, opt_input_file);
     if (ret < 0)
-        exit(ret == AVERROR_EXIT ? 0 : 1);
+        return NULL; //exit(ret == AVERROR_EXIT ? 0 : 1);
 
     if (!input_filename) {
         show_usage();
         av_log(NULL, AV_LOG_FATAL, "An input file must be specified\n");
         av_log(NULL, AV_LOG_FATAL,
                "Use -h to get full help or, even better, run 'man %s'\n", program_name);
-        exit(1);
+        return NULL; //exit(1);
     }
 
     if (display_disable) {
@@ -3804,7 +3830,7 @@ int main(int argc, char **argv)
     if (SDL_Init (flags)) {
         av_log(NULL, AV_LOG_FATAL, "Could not initialize SDL - %s\n", SDL_GetError());
         av_log(NULL, AV_LOG_FATAL, "(Did you set the DISPLAY variable?)\n");
-        exit(1);
+        return NULL; //exit(1);
     }
 
     SDL_EventState(SDL_SYSWMEVENT, SDL_IGNORE);
@@ -3846,6 +3872,7 @@ int main(int argc, char **argv)
         if (!window) {
             av_log(NULL, AV_LOG_FATAL, "Failed to create window: %s", SDL_GetError());
             do_exit(NULL);
+            return NULL;
         }
 
         if (vk_renderer) {
@@ -3856,6 +3883,7 @@ int main(int argc, char **argv)
                 if (ret < 0) {
                     av_log(NULL, AV_LOG_FATAL, "Failed to parse, %s\n", vulkan_params);
                     do_exit(NULL);
+                    return NULL;
                 }
             }
             ret = vk_renderer_create(vk_renderer, window, dict);
@@ -3863,6 +3891,7 @@ int main(int argc, char **argv)
             if (ret < 0) {
                 av_log(NULL, AV_LOG_FATAL, "Failed to create vulkan renderer, %s\n", av_err2str(ret));
                 do_exit(NULL);
+                return NULL;
             }
         } else {
             renderer = SDL_CreateRenderer(window, -1, SDL_RENDERER_ACCELERATED | SDL_RENDERER_PRESENTVSYNC);
@@ -3877,6 +3906,7 @@ int main(int argc, char **argv)
             if (!renderer || !renderer_info.num_texture_formats) {
                 av_log(NULL, AV_LOG_FATAL, "Failed to create window or renderer: %s", SDL_GetError());
                 do_exit(NULL);
+                return NULL;
             }
         }
     }
@@ -3885,11 +3915,70 @@ int main(int argc, char **argv)
     if (!is) {
         av_log(NULL, AV_LOG_FATAL, "Failed to initialize VideoState!\n");
         do_exit(NULL);
+        return NULL;
     }
 
-    event_loop(is);
-
-    /* never returns */
+    return is;
+}
 
-    return 0;
+// Global state reset (simplified for now, mimicking other libs)
+void ffplay_reset_internal_state(void) {
+    // Reset statics if possible or necessary.
+    // For ffplay, much state is in VideoState, but options are static.
+    // We can reset options via uninit_opts() which is called in do_exit.
+    // However, static ints like startup_volume, etc. persist.
+    // We should reset them:
+    input_filename = NULL;
+    window_title = NULL;
+    default_width  = 640;
+    default_height = 480;
+    screen_width  = 0;
+    screen_height = 0;
+    audio_disable = 0;
+    video_disable = 0;
+    subtitle_disable = 0;
+    seek_by_bytes = -1;
+    seek_interval = 10;
+    display_disable = 0;
+    borderless = 0;
+    alwaysontop = 0;
+    startup_volume = 100;
+    show_status = -1;
+    av_sync_type = AV_SYNC_AUDIO_MASTER;
+    start_time = AV_NOPTS_VALUE;
+    duration = AV_NOPTS_VALUE;
+    fast = 0;
+    genpts = 0;
+    lowres = 0;
+    decoder_reorder_pts = -1;
+    autoexit = 0;
+    exit_on_keydown = 0;
+    exit_on_mousedown = 0;
+    loop = 1;
+    framedrop = -1;
+    infinite_buffer = -1;
+    show_mode = SHOW_MODE_NONE;
+    audio_codec_name = NULL;
+    subtitle_codec_name = NULL;
+    video_codec_name = NULL;
+    rdftspeed = 0.02;
+    // cursor_last_shown // timestamp, doesn't matter much
+    cursor_hidden = 0;
+    // vfilters_list reset in do_exit
+    nb_vfilters = 0;
+    afilters = NULL;
+    autorotate = 1;
+    find_stream_info = 1;
+    filter_nbthreads = 0;
+    enable_vulkan = 0;
+    vulkan_params = NULL;
+    hwaccel = NULL;
+    
+    is_full_screen = 0;
+    
+    // window, renderer usually cleaned in do_exit/stream_close
+    window = NULL;
+    renderer = NULL;
+    vk_renderer = NULL;
 }
+
